src/core/lib/event_engine/posix_engine/ev_epoll1_linux.cc

```diff
@@ -76,57 +83,96 @@ class Epoll1EventHandle : public EventHandle {
     read_closure_->InitEvent();
     write_closure_->InitEvent();
     error_closure_->InitEvent();
-    pending_actions_ = 0;
+    pending_actions_.store(0, std::memory_order_relaxed);
+  }
+  void ReInit(int fd) {
+    fd_ = fd;
+    read_closure_->InitEvent();
+    write_closure_->InitEvent();
+    error_closure_->InitEvent();
+    pending_actions_.store(0, std::memory_order_relaxed);
   }
   Epoll1Poller* Poller() { return poller_; }
-  void SetPendingActions(bool pending_read, bool pending_write,
-                         bool pending_error) {
-    pending_actions_ |= pending_read;
+  EventEngine::Closure* SetPendingActions(bool pending_read, bool pending_write,
+                                          bool pending_error) {
+    int pending_actions = 0;
+    pending_actions |= pending_read;
     if (pending_write) {
-      pending_actions_ |= (1 << 2);
+      pending_actions |= (1 << 2);
     }
     if (pending_error) {
-      pending_actions_ |= (1 << 3);
+      pending_actions |= (1 << 3);
+    }
+    // Another thread may be executing ExecutePendingActions() at this point
+    // This is possible for instance, if one instantiation of Work(..) sets
+    // an fd to be readable while the next instantiation of Work(...) may
+    // set the fd to be writable. While the second instantiation is running,
+    // ExecutePendingActions() of the first instantiation may execute in
+    // parallel and read the pending_actions_ variable. So we need to use
+    // atomics to manipulate pending_actions_ variable.
+
+    // Load pending_actions_ into curr.
+    int curr = pending_actions_.load(std::memory_order_acquire);
+    while (true) {
+      // Check whether pending_actions_ == curr. If so set pending_actions_ to
+      // curr | pending_actions. Otherwise update curr to latest value of
+      // pending_actions_ and retry.
+      if (pending_actions_.compare_exchange_strong(curr, curr | pending_actions,
```

Comment:
Could we eliminate this by breaking our pending actions into three: read, write, and error - and having a different closure for each?

Reply:
Yes I used 3 atomic variables in the latest commit. It eliminates that loop.